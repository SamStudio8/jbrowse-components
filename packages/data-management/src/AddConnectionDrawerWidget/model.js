import { types } from 'mobx-state-tree'
import { ElementId } from '@gmod/jbrowse-core/mst-types'

export default types.model('AddConnectionModel', {
  id: ElementId,
  type: types.literal('AddConnectionDrawerWidget'),
})
