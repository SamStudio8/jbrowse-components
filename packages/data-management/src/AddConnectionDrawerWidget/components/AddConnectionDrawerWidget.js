import Button from '@material-ui/core/Button'
import Step from '@material-ui/core/Step'
import StepContent from '@material-ui/core/StepContent'
import StepLabel from '@material-ui/core/StepLabel'
import Stepper from '@material-ui/core/Stepper'
import { withStyles } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'
import { PropTypes as MobxPropTypes } from 'mobx-react'
import { observer } from 'mobx-react-lite'
import { getRoot } from 'mobx-state-tree'
import propTypes from 'prop-types'
import React, { useState } from 'react'
import ConfigureConnection from './ConfigureConnection'
import ConnectionTypeSelect from './ConnectionTypeSelect'

const styles = theme => ({
  root: {
    marginTop: theme.spacing.unit,
  },
  stepper: {
    backgroundColor: theme.palette.background.default,
  },
  button: {
    marginTop: theme.spacing.unit,
    marginRight: theme.spacing.unit,
  },
  actionsContainer: {
    marginBottom: theme.spacing.unit * 2,
  },
})

const steps = ['Select a Connection Type', 'Configure Connection']

function AddConnectionDrawerWidget(props) {
  const [connectionType, setConnectionType] = useState({})
  const [configModel, setConfigModel] = useState({})

  const [activeStep, setActiveStep] = useState(0)

  const { classes, model } = props
  const rootModel = getRoot(model)

  const { pluginManager } = rootModel

  function handleSetConnectionType(newConnectionType) {
    setConnectionType(newConnectionType)
    setConfigModel(newConnectionType.configSchema.create())
  }

  function stepContent() {
    switch (activeStep) {
      case 0:
        return (
          <ConnectionTypeSelect
            connectionTypeChoices={pluginManager.getElementTypesInGroup(
              'connection',
            )}
            connectionType={connectionType}
            setConnectionType={handleSetConnectionType}
          />
        )
      case 1:
        return (
          <ConfigureConnection
            connectionType={connectionType}
            model={configModel}
          />
        )

      default:
        return <Typography>Unknown step</Typography>
    }
  }

  function handleNext() {
    if (activeStep === steps.length - 1) handleFinish()
    else setActiveStep(activeStep + 1)
  }

  function handleBack() {
    setActiveStep(activeStep - 1)
  }

  function handleFinish() {
    rootModel.configuration.addConnection(configModel)
    rootModel.hideDrawerWidget(
      rootModel.drawerWidgets.get('addConnectionDrawerWidget'),
    )
  }

  function checkNextEnabled() {
    if (
      (activeStep === 0 && connectionType.name) ||
      (activeStep === 1 && configModel)
    )
      return true
    return false
  }

  return (
    <div className={classes.root}>
      <Stepper
        className={classes.stepper}
        activeStep={activeStep}
        orientation="vertical"
      >
        {steps.map(label => (
          <Step key={label}>
            <StepLabel>{label}</StepLabel>
            <StepContent>
              {stepContent()}
              <div className={classes.actionsContainer}>
                <Button
                  disabled={activeStep === 0}
                  onClick={handleBack}
                  className={classes.button}
                >
                  Back
                </Button>
                <Button
                  disabled={!checkNextEnabled()}
                  variant="contained"
                  color="primary"
                  onClick={handleNext}
                  className={classes.button}
                  data-testid="addConnectionNext"
                >
                  {activeStep === steps.length - 1 ? 'Connect' : 'Next'}
                </Button>
              </div>
            </StepContent>
          </Step>
        ))}
      </Stepper>
    </div>
  )
}

AddConnectionDrawerWidget.propTypes = {
  classes: propTypes.objectOf(propTypes.string).isRequired,
  model: MobxPropTypes.observableObject.isRequired,
}

export default withStyles(styles)(observer(AddConnectionDrawerWidget))
