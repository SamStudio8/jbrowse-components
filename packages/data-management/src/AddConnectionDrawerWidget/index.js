import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'

export const ReactComponent = import('./components/AddConnectionDrawerWidget')
export { default as stateModel } from './model'
export const configSchema = ConfigurationSchema('AddConnectionDrawerWidget', {})
