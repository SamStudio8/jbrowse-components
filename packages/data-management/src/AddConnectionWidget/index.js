import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'

export const ReactComponent = import('./components/AddConnectionWidget')
export { default as stateModel } from './model'
export const configSchema = ConfigurationSchema('AddConnectionWidget', {})
