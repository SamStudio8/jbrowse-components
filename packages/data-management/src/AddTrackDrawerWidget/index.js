import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'

export const ReactComponent = import('./components/AddTrackDrawerWidget')
export { default as stateModelFactory } from './model'
export const configSchema = ConfigurationSchema('AddTrackDrawerWidget', {})
