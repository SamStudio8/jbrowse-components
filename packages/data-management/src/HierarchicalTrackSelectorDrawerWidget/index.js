import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'

export const ReactComponent = import('./components/HierarchicalTrackSelector')
export { default as stateModelFactory } from './model'
export const configSchema = ConfigurationSchema(
  'HierarchicalTrackSelectorDrawerWidget',
  {},
)
