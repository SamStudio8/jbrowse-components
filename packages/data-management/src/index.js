import { lazy } from 'react'
import ConnectionType from '@gmod/jbrowse-core/pluggableElementTypes/ConnectionType'
import DrawerWidgetType from '@gmod/jbrowse-core/pluggableElementTypes/DrawerWidgetType'
import Plugin from '@gmod/jbrowse-core/Plugin'
import {
  configSchema as ucscConfigSchema,
  modelFactory as ucscModelFactory,
} from './ucsc-trackhub'
import {
  configSchema as jbrowse1ConfigSchema,
  modelFactory as jbrowse1ModelFactory,
} from './jbrowse1'
import {
  reactComponent as AddTrackReactComponent,
  stateModelFactory as AddTrackStateModelFactory,
  configSchema as AddTrackConfigSchema,
} from './AddTrackDrawerWidget'
import {
  reactComponent as AddConnectionReactComponent,
  stateModel as AddConnectionStateModel,
  configSchema as AddConnectionConfigSchema,
} from './AddConnectionDrawerWidget'
import {
  reactComponent as HierarchicalTrackSelectorReactComponent,
  stateModelFactory as HierarchicalTrackSelectorStateModelFactory,
  configSchema as HierarchicalTrackSelectorConfigSchema,
} from './HierarchicalTrackSelectorDrawerWidget'

export default class extends Plugin {
  install(pluginManager) {
    pluginManager.addConnectionType(
      () =>
        new ConnectionType({
          name: 'UCSCTrackHubConnection',
          configSchema: ucscConfigSchema,
          stateModel: ucscModelFactory(pluginManager),
          displayName: 'UCSC Track Hub',
          description: 'A track or assembly hub in the Track Hub format',
          url: '//genome.ucsc.edu/goldenPath/help/hgTrackHubHelp.html#Intro',
        }),
    )

    pluginManager.addConnectionType(
      () =>
        new ConnectionType({
          name: 'JBrowse1Connection',
          configSchema: jbrowse1ConfigSchema,
          stateModel: jbrowse1ModelFactory(pluginManager),
          displayName: 'JBrowse 1 Data',
          description: 'A JBrowse 1 data directory',
          url: '//jbrowse.org/',
        }),
    )

    pluginManager.addDrawerWidgetType(() => {
      return new DrawerWidgetType({
        name: 'HierarchicalTrackSelectorDrawerWidget',
        heading: 'Available Tracks',
        configSchema: HierarchicalTrackSelectorConfigSchema,
        stateModel: HierarchicalTrackSelectorStateModelFactory(pluginManager),
        LazyReactComponent: lazy(() => HierarchicalTrackSelectorReactComponent),
      })
    })

    pluginManager.addDrawerWidgetType(() => {
      return new DrawerWidgetType({
        name: 'AddTrackDrawerWidget',
        heading: 'Add a track',
        configSchema: AddTrackConfigSchema,
        stateModel: AddTrackStateModelFactory(pluginManager),
        LazyReactComponent: lazy(() => AddTrackReactComponent),
      })
    })

    pluginManager.addDrawerWidgetType(() => {
      return new DrawerWidgetType({
        name: 'AddConnectionDrawerWidget',
        heading: 'Add a connection',
        configSchema: AddConnectionConfigSchema,
        stateModel: AddConnectionStateModel,
        LazyReactComponent: lazy(() => AddConnectionReactComponent),
      })
    })
  }
}
