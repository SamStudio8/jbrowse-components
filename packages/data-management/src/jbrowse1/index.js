import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'

export { default as modelFactory } from './model'

export const configSchema = ConfigurationSchema(
  'JBrowse1Connection',
  {
    name: {
      type: 'string',
      defaultValue: 'nameOfJBrowse1Connection',
    },
    dataDirLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/to/data/dir/' },
    },
    assemblyName: {
      type: 'string',
      defaultValue: '',
    },
    useAssemblySequences: {
      type: 'boolean',
      defaultValue: false,
    },
  },
  { explicitlyTyped: true },
)
