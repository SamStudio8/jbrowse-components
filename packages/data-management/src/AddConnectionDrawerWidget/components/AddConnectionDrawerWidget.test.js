import {
  render,
  cleanup,
  fireEvent,
  // waitForElement,
} from 'react-testing-library'
import React from 'react'
import { createTestEnv } from '@gmod/jbrowse-web/src/JBrowse'
import AddConnectionDrawerWidget from './AddConnectionDrawerWidget'

describe('<AddConnectionDrawerWidget />', () => {
  let model

  beforeAll(async () => {
    const { rootModel } = await createTestEnv({
      configId: 'testing',
      defaultSession: {},
      rpc: { configId: 'testingRpc' },
    })
    rootModel.addDrawerWidget(
      'AddConnectionDrawerWidget',
      'addConnectionDrawerWidget',
    )
    model = rootModel.drawerWidgets.get('addConnectionDrawerWidget')
  })

  afterEach(cleanup)

  it('renders', () => {
    const { container } = render(<AddConnectionDrawerWidget model={model} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it.todo('Fix the tests below once react async testing stuff improves')

  xit('can handle a custom UCSC trackHub URL', async () => {
    //     const mockFetch = url => {
    //       const urlText = url.href ? url.href : url
    //       let responseText = ''
    //       if (urlText.endsWith('hub.txt'))
    //         responseText = `hub TestHub
    // shortLabel Test Hub
    // longLabel Test Genome Informatics Hub for human DNase and RNAseq data
    // genomesFile genomes.txt
    // email genome@test.com
    // descriptionUrl test.html
    // `
    //       else if (urlText.endsWith('genomes.txt'))
    //         responseText = `genome testAssembly
    // trackDb hg19/trackDb.txt
    // `
    //       return Promise.resolve(
    //         new Response(responseText, { url: 'http://test.com/hub.txt' }),
    //       )
    //     }
    //     jest.spyOn(global, 'fetch').mockImplementation(mockFetch)
    //     const { getByTestId /* , getByText */ } = render(
    //       <AddConnectionDrawerWidget model={model} />,
    //     )
    //     fireEvent.click(getByTestId('ucsc'))
    //     fireEvent.click(getByTestId('addConnectionNext'))
    //     fireEvent.click(getByTestId('ucscCustom'))
    //     fireEvent.click(getByTestId('addConnectionNext'))
    //     fireEvent.change(getByTestId('trackHubUrlInput'), {
    //       target: { value: 'http://test.com/hub.txt' },
    //     })
    //     fireEvent.click(getByTestId('trackHubUrlInputValidate'))
    // Next line doesn't work yet
    // await waitForElement(() => getByText('Assemblies'))
    // Do the rest of the UI actions to add the connection
  })

  xit('can handle Track Hub Registry hub', () => {})
})
