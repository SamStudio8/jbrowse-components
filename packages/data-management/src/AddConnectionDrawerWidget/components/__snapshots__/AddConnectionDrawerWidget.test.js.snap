// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddConnectionDrawerWidget /> renders 1`] = `
<div
  class="AddConnectionDrawerWidget-root-1"
>
  <div
    class="MuiPaper-root-9 MuiPaper-elevation0-11 MuiStepper-root-5 MuiStepper-vertical-7 AddConnectionDrawerWidget-stepper-2"
  >
    <div
      class="MuiStep-root-36 MuiStep-vertical-38"
    >
      <span
        class="MuiStepLabel-root-41 MuiStepLabel-vertical-43"
      >
        <span
          class="MuiStepLabel-iconContainer-49"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-57 MuiStepIcon-root-52 MuiStepIcon-active-54"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <circle
              cx="12"
              cy="12"
              r="12"
            />
            <text
              class="MuiStepIcon-text-53"
              text-anchor="middle"
              x="12"
              y="16"
            >
              1
            </text>
          </svg>
        </span>
        <span
          class="MuiStepLabel-labelContainer-51"
        >
          <span
            class="MuiTypography-root-66 MuiTypography-body1-75 MuiStepLabel-label-44 MuiStepLabel-active-45"
          >
            Select a Connection Type
          </span>
        </span>
      </span>
      <div
        class="MuiStepContent-root-102"
        icon="1"
      >
        <div
          class="MuiCollapse-container-105 MuiCollapse-entered-106 MuiStepContent-transition-104"
          style="min-height: 0px;"
        >
          <div
            class="MuiCollapse-wrapper-107"
          >
            <div
              class="MuiCollapse-wrapperInner-108"
            >
              <form
                autocomplete="off"
              >
                <div
                  class="MuiFormControl-root-109 MuiFormControl-fullWidth-112"
                >
                  <div
                    class="MuiInputBase-root-133 MuiInput-root-120 MuiInput-underline-124 MuiInputBase-formControl-134 MuiInput-formControl-121"
                  >
                    <div
                      class="MuiSelect-root-113"
                    >
                      <div
                        aria-haspopup="true"
                        aria-pressed="false"
                        class="MuiSelect-select-114 MuiSelect-selectMenu-117 MuiInputBase-input-143 MuiInput-input-128"
                        role="button"
                        tabindex="0"
                      >
                        <span>
                          â€‹
                        </span>
                      </div>
                      <input
                        type="hidden"
                        value=""
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root-57 MuiSelect-icon-119"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M7 10l5 5 5-5z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </form>
              <div
                class="AddConnectionDrawerWidget-actionsContainer-4"
              >
                <button
                  class="MuiButtonBase-root-180 MuiButtonBase-disabled-181 MuiButton-root-154 MuiButton-text-156 MuiButton-flat-159 MuiButton-disabled-174 AddConnectionDrawerWidget-button-3"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <span
                    class="MuiButton-label-155"
                  >
                    Back
                  </span>
                </button>
                <button
                  class="MuiButtonBase-root-180 MuiButtonBase-disabled-181 MuiButton-root-154 MuiButton-contained-165 MuiButton-containedPrimary-166 MuiButton-raised-168 MuiButton-raisedPrimary-169 MuiButton-disabled-174 AddConnectionDrawerWidget-button-3"
                  data-testid="addConnectionNext"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <span
                    class="MuiButton-label-155"
                  >
                    Next
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiStepConnector-root-183 MuiStepConnector-vertical-185 MuiStepConnector-disabled-189"
    >
      <span
        class="MuiStepConnector-line-190 MuiStepConnector-lineVertical-192"
      />
    </div>
    <div
      class="MuiStep-root-36 MuiStep-vertical-38"
    >
      <span
        class="MuiStepLabel-root-41 MuiStepLabel-vertical-43 MuiStepLabel-disabled-48"
      >
        <span
          class="MuiStepLabel-iconContainer-49"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-57 MuiStepIcon-root-52"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <circle
              cx="12"
              cy="12"
              r="12"
            />
            <text
              class="MuiStepIcon-text-53"
              text-anchor="middle"
              x="12"
              y="16"
            >
              2
            </text>
          </svg>
        </span>
        <span
          class="MuiStepLabel-labelContainer-51"
        >
          <span
            class="MuiTypography-root-66 MuiTypography-body1-75 MuiStepLabel-label-44"
          >
            Configure Connection
          </span>
        </span>
      </span>
      <div
        class="MuiStepContent-root-102 MuiStepContent-last-103"
        disabled=""
        icon="2"
      />
    </div>
  </div>
</div>
`;
